### PR TITLE
Layer comprehensive demo

### PR DESCRIPTION
**Pull request by buraklayer**

### CODE REVIEW

**Comment by mwitiderrick** [comment]
Very cool commit messages ;)

**Comment by mwitiderrick** [comment]
@buraklayer can we add this to https://github.com/layerai/examples/tree/main/tutorials? wdyt?

**Comment by mwitiderrick** [comment]
In the readme, check that open in Layer and open in colab are properly linked

**Comment by mwitiderrick** [comment]
@mecevit I think we need to decide which English we will use in all our communication for consistency, American or Britain. optimise and optimize are the same, but I think it would be useful to settle on a standard for all communication.

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb [hunk 1]
```diff
@@ -0,0 +1,868 @@
...
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "source": [
+        "# EXPERIMENT TRACKING FOR ML PROJECTS\n",
+        "\n",
+        "In this tutorial, we will discuss what experiment tracking means within the Machine Learning context, why it is important to keep track of your project journey and how you could make use of Layer platform to do it.<br><br>\n",
+        "ML project development is experimental by its nature. It means that we need to run many recurrent experiments to end up with the best performing model. These experiments might include various tasks such as parameter searching to optimise a model or transforming a dataset in an incremental manner. You could think of these experiments as your Python functions in your notebook designed for conducting a specific task.<br><br>\n",
+        "\n",
+        "***Experiment Tracking:***<br> \n",
```
Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:27]: Can we add the items below in a list?

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb [hunk 2]
```diff
@@ -0,0 +1,868 @@
...
+      "name": "python"
+    }
+  },
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "source": [
+        "# EXPERIMENT TRACKING FOR ML PROJECTS\n",
+        "\n",
+        "In this tutorial, we will discuss what experiment tracking means within the Machine Learning context, why it is important to keep track of your project journey and how you could make use of Layer platform to do it.<br><br>\n",
```
Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:24]: Link experiment tracking to layer.ai

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb [hunk 3]
```diff
@@ -0,0 +1,868 @@
...
+        "ML project development is experimental by its nature. It means that we need to run many recurrent experiments to end up with the best performing model. These experiments might include various tasks such as parameter searching to optimise a model or transforming a dataset in an incremental manner. You could think of these experiments as your Python functions in your notebook designed for conducting a specific task.<br><br>\n",
+        "\n",
+        "***Experiment Tracking:***<br> \n",
+        "*\\\"During development of an ML project, being able to save entire project history and thus revert to any previous project snapshot when needed.\\\"*<br><br>\n",
+        "Let's list possible actions data scientists should take during an ML project development and experiment tracking:<br><br>\n",
+        "**Versioning:** As Github is doing for codes, datasets and models should be versioned automatically each time you build them so that whenever needed entities could be reverted to any previous version. It is sort of applying the CI/CD software principle for data projects as well.<br> \n",
```
Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:30]: GitHub

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb [hunk 4]
```diff
@@ -0,0 +1,868 @@
...
+        "**Monitoring:** Display some information about datasets and models such as model performance metrics, model parameters or data profiling. It will provide continuous observability on entities and comparing different versions. \n",
+        "Testing: Sanity check on datasets and models by running simple unit tests to have more reliable pipeline. It could be perceived as an Alert System for data projects.<br>\n",
+        "**Documentation:** Writing descriptions about datasets and models as well as a detailed project report. It will tell other people what's a project is all about and break work in-silos.<br>\n",
```
Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:33]: what's to what

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb [hunk 5]
```diff
@@ -0,0 +1,868 @@
...
+        "id": "IUWbLf-99xW1"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "# **Step I:** Install Layer and Authentication\n",
+        "---\n",
+        "\n",
+        "With just 3 lines of code, you will be done with installation and authentication in your notebook. .<br><br>\n",
```
Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:95]: Remove extra dot after notebook

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb [hunk 6]
```diff
@@ -0,0 +1,868 @@
...
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "\n",
+        "### Layer Decorators or Functions Used In Step-II\n",
+        "---\n",
+        "\n",
+        "**layer.init():**<br>\n",
+        "To start logging into the ***my-first-project*** project, layer's init function should be called with the name of project as shown in the first line below. It basically tells Layer that everything will be saved in that specific project. If any project exists with the given name, then it will match and sync otherwise, a new project with the given name will be created under your organisation or personal account. This function also has a ***pip_packages*** parameter which takes a list of python packages. You are expected to list all Layer packages you use in your project. Layer will install those python packages first to make your project environment ready on the Layer Infra. <br><br>\n",
```
Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:171]: Layer's init

Comment (mwitiderrick) [reply comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:171]: Python packages

Comment (mwitiderrick) [reply comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:171]: `You are expected to list all Layer packages you use in your project` is not entirely true, we do have some packages already installed. They should only specify the pages if the versions we have are different from the ones they want to use https://docs.app.layer.ai/docs/reference/fabrics#preinstalled-libraries

Comment (mwitiderrick) [reply comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:171]: If this is for demonstration purposes, we should add the package versions.

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb [hunk 7]
```diff
@@ -0,0 +1,868 @@
...
+        "\n",
+        "### Layer Decorators and Functions Used in Step-III\n",
+        "\n",
+        "---\n",
+        "\n",
+        "\n",
+        "\n",
+        "\n",
+        "**@dataset:**<br> \n",
+        "It is a decorator which defines the pandas data frame ***training_df*** returned by the function as a Layer dataset named ***my_first_dataset*** under the project ***my_first_project***. Layer starts tracking this dataset automatically every time you run this function. In other words, Layer will do versioning on your dataset and log any other data you want Layer to store along with this dataset. It is important to note that it is required that you could use this dataset decorator at the top a Python function returning a pandas data frame.<br><br>\n",
```
Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:229]: Pandas -- we can link to the `to_pandas` function.

Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:229]: `It is important to note that it is required that you could use this dataset decorator at the top a Python function returning a Pandas data frame.` Is `could` necessary here?

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb [hunk 8]
```diff
@@ -0,0 +1,868 @@
...
+        "https://docs.app.layer.ai/docs/sdk-library/assertions<br><br>\n",
+        "\n",
+        "**@resources:**<br> \n",
+        "It's a decorator that is used to upload local files to Layer remote machines. In this example, it is used to upload files under the ***data*** directory on Google Colab file system to the Layer Cloud. Folder structure will be exactly the same on the cloud side as your local so that you don't have to change any hard-coded file paths exist in your code in order to run your function remotely.<br><br> \n",
+        "For more information about the resources decorator:<br>\n",
+        "https://docs.app.layer.ai/docs/sdk-library/resources-decorator<br><br>\n",
+        "\n",
+        "\n",
+        "**layer.log():**<br>\n",
+        "In the function body, there is also one more layer supported function which is used to log arbitrary data along with the dataset. This function supports many different data types such as primitives, dataframes or images. In this example, we log a box plot of the column named ***target***. In general, you could log anything attached with the current build version of your dataset or model by using this function.\n",
```
Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:245]: Layer supported function

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb [hunk 9]
```diff
@@ -0,0 +1,868 @@
...
+      "metadata": {
+        "id": "c8vKAtvLzZH7"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Now, we are ready to take one more step to run the function above on Layer Cloud remotely. <br>\n",
```
Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:300]: remotely might not be necessary here.

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb [hunk 10]
```diff
@@ -0,0 +1,868 @@
...
+        "\n",
+        "When you run the cell below for the first time, it will create your dataset ***my_first_dataset v1.1*** and also print a link to that dataset's page on Layer Web UI. Once you click on the generated link, you will see a dataset page looks like this:\n",
+        "\n",
```
Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:303]: you will see a dataset page that looks like this

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb [hunk 11]
```diff
@@ -0,0 +1,868 @@
...
+      "metadata": {
+        "id": "crHeFPUVLryD"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Now, let's make some changes in the function body that affects the returning dataset's schema such as adding one more column. Assume that you have come up with a new feature that you think it will help improving your ML model's prediction performance. What normally you would do in the next iteration (or experiment) is to add this new feature into your dataset and train your model using the new version of your dataset. In the case below, we will simulate it by increasing number of features ***n_features*** in the ***make_regression*** function from 3 to 4.<br>\n",
```
Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:332]: you think it will help improve your ML model's prediction performance

Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:332]: What you would normally do...

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb [hunk 12]
```diff
@@ -0,0 +1,868 @@
...
+        "\n",
+        "You could see more about the assert_true decorator here:<br>\n",
+        "https://docs.app.layer.ai/docs/sdk-library/assert-true<br><br>\n",
+        "\n",
+        "**get_dataset():**<br>\n",
+        "This function retrieves a Layer dataset object from the Datasets you defined on Layer previously. It returns a Layer Dataset object and you should also call ***to_pandas()*** function to convert it to a pandas data frame. In the case below, we fetches ***my_first_dataset v1.2*** to use as a training data to build the model.<br><br>\n",
```
Comment (mwitiderrick) [inline comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:442]: to convert it to a Pandas data frame

**Review by mwitiderrick — COMMENTED** [review]
If users are expected to run this notebook we need to remove `burakozen/my-first-project/` your usernames from the notebook, otherwise, we should let them know how to change this. 

```
@model("my_first_model" , dependencies=[Dataset('burakozen/my-first-project/datasets/my_first_dataset')])

```

**Comment by mwitiderrick** [comment]
Where you talk about the readme, maybe we can link it so that they can also try to upload it with Layer

**Comment by buraklayer** [comment]
> @buraklayer can we add this to https://github.com/layerai/examples/tree/main/tutorials? wdyt?

Yes we can. Maybe, under a new folder such as 'comprehensive_demos' since we will have more of this in the upcoming days.

**Comment by buraklayer** [comment]
> @mecevit I think we need to decide which English we will use in all our communication for consistency, American or Britain. optimise and optimize are the same, but I think it would be useful to settle on a standard for all communication.

Not a big deal, let's go with the American English then.

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb (hunk 1 above)
Comment (buraklayer) [reply comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:27]: > Can we add the items below in a list?

Done

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb (hunk 6 above)
Comment (buraklayer) [reply comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:171]: Done

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb (hunk 7 above)
Comment (buraklayer) [reply comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:229]: removed

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb (hunk 9 above)
Comment (buraklayer) [reply comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:300]: removed

File: comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb (hunk 11 above)
Comment (buraklayer) [reply comprehensive-getting-started/Layer_Comprehensive_Getting_Started.ipynb:332]: No, but better to sketch a scenario here.

**Pull request closed without merging**